source/http_server.c

```diff
@@ -0,0 +1,230 @@
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_incoming_connection)) {
+        PyErr_SetString(PyExc_ValueError, "on_incoming_connection callback is required");
+        goto error;
+    }
+
+    if (!PyCallable_Check(on_destroy_complete)) {
+        PyErr_SetString(PyExc_TypeError, "on_destroy_complete is invalid");
+        goto error;
+    }
+
+    struct aws_server_bootstrap *bootstrap = PyCapsule_GetPointer(bootstrap_capsule, s_capsule_name_server_bootstrap);
+    if (!bootstrap) {
+        goto error;
+    }
+
+    struct aws_tls_connection_options *connection_options = NULL;
+
+    if (tls_conn_options_capsule != Py_None) {
+        connection_options = PyCapsule_GetPointer(tls_conn_options_capsule, s_capsule_name_tls_conn_options);
+        if(!connection_options){
+            goto error;
+        }
+    }
+
+    py_server = aws_mem_calloc(allocator, 1, sizeof(struct py_http_server));
+    if (!py_server) {
+        PyErr_SetAwsLastError();
+        goto error;
+    }
+
+    struct aws_socket_options socket_options;
+    if (!aws_socket_options_init_from_py(&socket_options, py_socket_options)) {
+        goto error;
+    }
+
+    py_server->on_incoming_connection = on_incoming_connection;
+
+    
+    py_server->on_destroy_complete = on_destroy_complete;
+    
+    py_server->allocator = allocator;
+
+    struct aws_http_server_options options;
+    AWS_ZERO_STRUCT(options);
+    options.self_size = sizeof(options);
+    options.bootstrap = bootstrap;
+    options.tls_options = connection_options;
+    options.allocator = allocator;
+    options.server_user_data = py_server;
+    struct aws_socket_endpoint endpoint;
+    AWS_ZERO_STRUCT(endpoint);
+
+    snprintf(endpoint.address, host_name_len, "%s", host_name);
+    endpoint.port = port_number;
+    options.socket_options = &socket_options;
+    options.endpoint = &endpoint;
+    options.on_incoming_connection = s_on_incoming_connection;
+    options.on_destroy_complete = s_on_destroy_complete;
+    PyObject *capsule = NULL;
+    py_server->server = aws_http_server_new(&options);
+    
+    if (py_server->server) {
+        /* success */
+        capsule = PyCapsule_New(py_server, s_capsule_name_http_server, s_http_server_destructor);
+        if(!capsule){
+            goto error;
+        }
+        py_server->capsule = capsule;
+        Py_INCREF(on_incoming_connection);
+        Py_INCREF(on_destroy_complete);
+        return capsule;
+    }
+
+error:
+    if (py_server) {
+        aws_mem_release(py_server->allocator, py_server);
+    }
+    return NULL;
+}
+
+PyObject *aws_py_http_server_release(PyObject *self, PyObject *args) {
+    (void)self;
+
+    PyObject *server_capsule = NULL;
+
+    if (PyArg_ParseTuple(args, "O", &server_capsule)) {
+        if (server_capsule != Py_None) {
+            struct py_http_server *py_server = PyCapsule_GetPointer(server_capsule, s_capsule_name_http_server);
+            if(!py_server){
+                Py_RETURN_NONE;
```

Comment:
In this case, shouldn't we `return NULL;` and cause an exception to be raised in python-land?

Reply:
I made this change to request handler branch, but forget to update here...